apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,590 @@
+            tracing::debug!(%span.name, ?span.start_time, ?span.end_time);
+            tracing::debug!("span: {:?}", span);
+            if let Some(query) = span
+                .attributes
+                .get(&opentelemetry::Key::from_static_str("query"))
+            {
+                tracing::debug!("query: {}", query);
+                let not_found = Value::String(Cow::from("not found"));
+                let client_name = span
+                    .attributes
+                    .get(&opentelemetry::Key::from_static_str("client_name"))
+                    .unwrap_or(&not_found)
+                    .to_string();
+                let client_version = span
+                    .attributes
+                    .get(&opentelemetry::Key::from_static_str("client_version"))
+                    .unwrap_or(&not_found)
+                    .to_string();
+                let operation_name = span
+                    .attributes
+                    .get(&opentelemetry::Key::from_static_str("operation_name"));
+                // XXX Since normalization is expensive, try to reduce the
+                // amount of normalization by doing an exact string match
+                // on a query. This might not save a lot of work and may
+                // result in too much caching, so re-visit this decision
+                // post-integration.
+                let key = self
+                    .normalized_queries
+                    .entry(query.as_str().to_string())
+                    .or_insert_with(|| normalize(operation_name, &query.as_str()));
+
+                // Retrieve DurationHistogram from our HashMap, or add a new one
+                let dh = dh_map
+                    .entry((client_name, client_version, key.clone()))
+                    .or_insert_with(|| DurationHistogram::new(None));
+                dh.increment_duration(span.end_time.duration_since(span.start_time).unwrap(), 1);
+            }
+        }
+
+        // Guarantee that the reporter is initialised
+        self.reporter
+            .get_or_try_init(|| async {
+                Reporter::try_new(self.collector.clone())
+                    .await
+                    .map_err::<ApolloError, _>(Into::into)
+            })
+            .await?;
+        let reporter = self.reporter.get_mut().unwrap();
+        // Convert the configuration data into a reportable form
+        let graph: ReporterGraph = self.graph.as_ref().unwrap().into();
+
+        // Report our consolidated statistics
+        for ((client_name, client_version, key), dh) in dh_map.into_iter() {
+            tracing::debug!("reporting entries: {}", dh.entries);
+            let stats = ContextualizedStats {
+                context: Some(StatsContext {
+                    client_name,
+                    client_version,
+                }),
+                query_latency_stats: Some(QueryLatencyStats {
+                    latency_count: dh.buckets,
```

Comment:
This doesn't use the run-length encoding that replaces bunches of 0s with negative numbers.  That should at least be called out via comment or something.

Reply:
I'll file an issue for considering future implementation. Spending too much time in the export() fn exerts back-pressure into the telemetry framework and leads to dropped data spans. Given that the data is ultimately gzipped for transfer, I wasn't entirely sure that this extra processing (in the hot path) was worth performing. 

I'll raise an issue to capture this thinking in a comment.

Reply:
That's an interesting point — I'm not sure we've done any real benchmarking of whether the run-length encoding adds much when using gzip. (Certainly more memory will be used temporarily in building the protobuf but maybe that doesn't matter.)